### PR TITLE
Optimize icon assets #2082

### LINKED ISSUE #2082: Optimize font assets loaded for icons
### Please confirm that you have searched existing issues in the repo

Yes, I have searched the existing issues

### Any related issues?

Very vaguely related https://github.com/MarkBind/markbind/issues/1670

### What is the area that this feature belongs to?

Performance

### Is your feature request related to a problem? Please describe.

We load all font assets for [icons](https://markbind.org/userGuide/formattingContents.html#icons) regardless of whether they are actually used in the page.

This can be included on an as-needed basis isntead.

### Describe the solution you'd like

The rough idea would be to:
1. Detect if these icons have been used in the page or layout.
2. Include the asset link only as necessary

Before this, investigate if conditionally including these asset files would cause side effects as well. E.g. does the octicon asset file include some global styles, that when removed, would cause things to look or behave different?

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jmestxr**
**What is the purpose of this pull request?**

Exclude icon asset CSS files (i.e. font awesome, material-icons, glyphicons, octicons) from being loaded if they are not used in the page

- [ ] Documentation update
- [ ] Bug fix
- [x] Feature addition or enhancement
- [ ] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

Resolves #2082

**Overview of changes:**

Implemented `filterIconAssets(preVueSsrHtml: string, postVueSsrHtml: string)` where:
- cheerio is used to check if the html of the page (including its layout, e.g. navbar, site-nav) contains elements that use icon CSS classes
- If no element using the CSS class from an icon library, then remove the CSS asset file of that library

**Anything you'd like to highlight/discuss:**

**Testing instructions:**

1. cd into `packages/cli/test/functional/test_site`. Run `markbind serve -d`.
1. To test page content with icons, test the pages: `testGlyphiconInPage.html`, `testFontAwesomeInPage.html`, `testMaterialIconsInPage.html`, `testOcticonInPage.html`.
1. To test site layout with icons, test the page `testIconsInSiteLayout.html`.

**Proposed commit message: (wrap lines at 72 characters)**
Optimize loading of icon asset files

---

**Checklist:** :ballot_box_with_check:

- [ ] Updated the documentation for feature additions and enhancements
- [x] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes

### CODE REVIEW

**Comment by jmestxr** [comment]
I'm not sure why the cheerio is not detecting icons in the layout (sitenav/navbar...). For example, when running the following code in Markbind:

```
const $ = cheerio.load(`<div><header>
<navbar type="dark" default-highlight-on="sibling-or-child">
  <template #brand><a href="/" title="Home" class="navbar-brand">MarkBind Test Site</a></template>
  <li><a class="nav-link" href="/test_site/bugs/index.html">Open Bugs</a></li></navbar>
<div class="bg-info display-4 text-center text-white">
    <br>
    Test Jumbotron<br>
    <br></div></header>
<p><strong>Relative Link Test</strong> This is a relative Intra-Site link in a layout (see <a href="/test_site/index.html#heading-with-hidden-keyword">link</a>)</p></div>
<div id="flex-body">
<overlay-source id="site-nav" tag-name="nav" to="site-nav">
  <div class="site-nav-top">
    <div class="fw-bold mb-2" style="font-size: 1.25rem;"><div><h2 id="default-layout">Default Layout<a class="fa fa-anchor" href="#default-layout" onclick="event.stopPropagation()"></a></h2></div></div></div>
  <div class="nav-component slim-scroll">
    <div><site-nav><overlay-source class="site-nav-list site-nav-list-root" tag-name="ul" to="mb-site-nav">
<li><div class="site-nav-default-list-item site-nav-list-item-0" onclick="handleSiteNavClick(this)"><a href="/test_site/index.html">Home 🏠</a></div></li>
<li><div class="site-nav-default-list-item site-nav-list-item-0" onclick="handleSiteNavClick(this)"><a href="/test_site/bugs/index.html">Open Bugs 🐛</a></div></li>
<li class="site-nav-custom-list-item site-nav-list-item-0">
<h3 id="testing-site-nav">Testing Site-Nav<a class="fa fa-anchor" href="#testing-site-nav" onclick="event.stopPropagation()"></a></h3>
</li>
<li><div class="site-nav-default-list-item site-nav-list-item-0" onclick="handleSiteNavClick(this)"><strong>Dropdown </strong> <span aria-hidden="true" class="glyphicon glyphicon-search"></span> title ✏️ <!-- Nested list items will be placed inside a Dropdown menu -->

<div class="site-nav-dropdown-btn-container"><i class="site-nav-dropdown-btn-icon site-nav-rotate-icon" onclick="handleSiteNavClick(this.parentNode.parentNode, false); event.stopPropagation();">
<span class="glyphicon glyphicon-menu-down" aria-hidden="true"></span>
</i></div></div><ul class="site-nav-dropdown-container site-nav-dropdown-container-open site-nav-list">
<li><div class="site-nav-default-list-item site-nav-list-item-1" onclick="handleSiteNavClick(this)"><a href="https://www.google.com/">Dropdown link one</a></div></li>
<li><div class="site-nav-default-list-item site-nav-list-item-1" onclick="handleSiteNavClick(this)"><a href="https://www.google.com/">Html within site-nav <span style="color: red;">should</span> be displayed properly</a></div></li>
<li><div class="site-nav-default-list-item site-nav-list-item-1" onclick="handleSiteNavClick(this)">Nested Dropdown title 📐

<div class="site-nav-dropdown-btn-container"><i class="site-nav-dropdown-btn-icon" onclick="handleSiteNavClick(this.parentNode.parentNode, false); event.stopPropagation();">
<span class="glyphicon glyphicon-menu-down" aria-hidden="true"></span>
</i></div></div><ul class="site-nav-dropdown-container site-nav-list">
<li><div class="site-nav-default-list-item site-nav-list-item-2" onclick="handleSiteNavClick(this)"><a href="https://www.google.com/"><strong>Nested</strong> Dropdown link one</a></div></li>
<li><div class="site-nav-default-list-item site-nav-list-item-2" onclick="handleSiteNavClick(this)"><a href="https://www.google.com/"><strong>Nested</strong> Dropdown link two</a></div></li>
</ul></li>
<li><div class="site-nav-default-list-item site-nav-list-item-1" onclick="handleSiteNavClick(this)"><a href="https://www.google.com/">Dropdown link two</a></div></li>
</ul></li>
<li><div class="site-nav-default-list-item site-nav-list-item-0" onclick="handleSiteNavClick(this)"><a href="https://www.google.com/"><mark>Third Link</mark> 📋</a></div></li>
<li><div class="site-nav-default-list-item site-nav-list-item-0" onclick="handleSiteNavClick(this)">Filler text <a href="https://www.youtube.com/"><span aria-hidden="true" class="glyphicon glyphicon-facetime-video"></span> Youtube 📺</a> filler text<!-- Using a link as a Dropdown title will not render a Dropdown menu -->

<div class="site-nav-dropdown-btn-container"><i class="site-nav-dropdown-btn-icon" onclick="handleSiteNavClick(this.parentNode.parentNode, false); event.stopPropagation();">
<span class="glyphicon glyphicon-menu-down" aria-hidden="true"></span>
</i></div></div><ul class="site-nav-dropdown-container site-nav-list">
<li><div class="site-nav-default-list-item site-nav-list-item-1" onclick="handleSiteNavClick(this)"><a href="https://www.youtube.com/watch?v=dQw4w9WgXcQ">The answer to everything in the universe</a></div></li>
<li><div class="site-nav-default-list-item site-nav-list-item-1" onclick="handleSiteNavClick(this)"><mark>Dropdown title</mark> <span aria-hidden="true" class="glyphicon glyphicon-comment"></span> ✏️ <!-- Dropdown menus in are still supported inside, as long as the title is not a link -->

<div class="site-nav-dropdown-btn-container"><i class="site-nav-dropdown-btn-icon site-nav-rotate-icon" onclick="handleSiteNavClick(this.parentNode.parentNode, false); event.stopPropagation();">
<span class="glyphicon glyphicon-menu-down" aria-hidden="true"></span>
</i></div></div><ul class="site-nav-dropdown-container site-nav-dropdown-container-open site-nav-list">
<li><div class="site-nav-default-list-item site-nav-list-item-2" onclick="handleSiteNavClick(this)"><a href="https://www.google.com/"><strong>Nested</strong> Dropdown link one</a></div></li>
</ul></li>
</ul></li>
<li><div class="site-nav-default-list-item site-nav-list-item-0" onclick="handleSiteNavClick(this)">Really Long Dropdown Title Really Long Dropdown Title Really Long Dropdown Title Really Long Dropdown

<div class="site-nav-dropdown-btn-container"><i class="site-nav-dropdown-btn-icon" onclick="handleSiteNavClick(this.parentNode.parentNode, false); event.stopPropagation();">
<span class="glyphicon glyphicon-menu-down" aria-hidden="true"></span>
</i></div></div><ul class="site-nav-dropdown-container site-nav-list">
<li class="site-nav-custom-list-item site-nav-list-item-1">Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text Really Really Long Text</li>
<li><div class="site-nav-default-list-item site-nav-list-item-1" onclick="handleSiteNavClick(this)">Nested Dropdown Title

<div class="site-nav-dropdown-btn-container"><i class="site-nav-dropdown-btn-icon" onclick="handleSiteNavClick(this.parentNode.parentNode, false); event.stopPropagation();">
<span class="glyphicon glyphicon-menu-down" aria-hidden="true"></span>
</i></div></div><ul class="site-nav-dropdown-container site-nav-list">
<li class="site-nav-custom-list-item site-nav-list-item-2">Hello Doge Hello Doge 🐶</li>
<li><div class="site-nav-default-list-item site-nav-list-item-2" onclick="handleSiteNavClick(this)"><a href="/test_site/index.html"><strong>NESTED LINK</strong> Home 🏠</a></div></li>
<li class="site-nav-custom-list-item site-nav-list-item-2">Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit Text cut off from height limit</li>
</ul></li>
</ul></li>
<li><div class="site-nav-default-list-item site-nav-list-item-0" onclick="handleSiteNavClick(this)">Test line break in navigation layout

<div class="site-nav-dropdown-btn-container"><i class="site-nav-dropdown-btn-icon" onclick="handleSiteNavClick(this.parentNode.parentNode, false); event.stopPropagation();">
<span class="glyphicon glyphicon-menu-down" aria-hidden="true"></span>
</i></div></div><ul class="site-nav-dropdown-container site-nav-list">
<li class="site-nav-custom-list-item site-nav-list-item-1">Nested line break text ✂️</li>
<li><div class="site-nav-default-list-item site-nav-list-item-1" onclick="handleSiteNavClick(this)"><a href="/test_site/index.html">Nested line break href</a>

<div class="site-nav-dropdown-btn-container"><i class="site-nav-dropdown-btn-icon" onclick="handleSiteNavClick(this.parentNode.parentNode, false); event.stopPropagation();">
<span class="glyphicon glyphicon-menu-down" aria-hidden="true"></span>
</i></div></div><ul class="site-nav-dropdown-container site-nav-list">
<li class="site-nav-custom-list-item site-nav-list-item-2">Nested Nested line break text ✂️</li>
</ul></li>
<li><div class="site-nav-default-list-item site-nav-list-item-1" onclick="handleSiteNavClick(this)">Nested line break dropdown menu

<div class="site-nav-dropdown-btn-container"><i class="site-nav-dropdown-btn-icon" onclick="handleSiteNavClick(this.parentNode.parentNode, false); event.stopPropagation();">
<span class="glyphicon glyphicon-menu-down" aria-hidden="true"></span>
</i></div></div><ul class="site-nav-dropdown-container site-nav-list">
<li class="site-nav-custom-list-item site-nav-list-item-2">Line break item 2 📘</li>
</ul></li>
</ul></li>
</overlay-source>
</site-nav></div></div></overlay-source>
<div id="content-wrapper">
  <breadcrumb></breadcrumb>
  
<div class="website-content">
<p><strong>Bug Description</strong></p>
<p><a href="/test_site/bugs/LINK_TO_MARKBIND_ISSUE">Issue #XX</a></p>
<p>Repro:</p>
<p>Steps to Reproduce</p></div>
</div>
<overlay-source id="page-nav" tag-name="nav" to="page-nav">
  <div class="nav-component slim-scroll">
    </div></overlay-source>    
<scroll-top-button></scroll-top-button></div>
<div><footer>
<h1 id="heading-in-footer-should-not-be-indexed">Heading in footer should not be indexed<a class="fa fa-anchor" href="#heading-in-footer-should-not-be-indexed" onclick="event.stopPropagation()"></a></h1>
<div class="text-center">
  This is a dynamic height footer that supports markdown <span>😄</span>!
</div></footer></div>
`);

console.log($("[class^=glyphicon]").length);

```

I got 0, even though there are glyphicon icons present in the sitenav of the above html.

On the contrary, when running the same code in a sandbox (also using cheerio v0.22.0 which MarkBind uses): https://codesandbox.io/s/cheerio-playground-forked-3lpdyc?file=/src/index.js:56-10907, i got 12.

Hmm, anyone has any ideas why this is happening? @MarkBind/active-devs

**Comment by jmestxr** [comment]
Currently there are some issues faced in cheerio parsing. The `content` in `filterIconAssets` is not fully processed as html and hence there are some non-html tags. cheerio is unable to parse these tags and hence it can't detect icons used within those tags. E.g. I have glyphicon icons used within <site-nav> tags which cheerio can't detect.

**Comment by jmestxr** [comment]
> Currently there are some issues faced in cheerio parsing. The `content` in `filterIconAssets` is not fully processed as html and hence there are some non-html tags. cheerio is unable to parse these tags and hence it can't detect icons used within those tags. E.g. I have glyphicon icons used within tags which cheerio can't detect.
> 
> Any ideas on how to do so? 🤔

It's okay, this is resolved

**Comment by tlylt** [comment]
@jmestxr
1. Any reason why we are creating a new test site for this?
2. Can we fix the test case error?
3. Does this work well during development? i.e when I'm writing content for a page, and I delete/include icons, will the page reload properly, and will there be any performance issue?

**Comment by jmestxr** [comment]
> @jmestxr
> 
> 1. Any reason why we are creating a new test site for this?
> 2. Can we fix the test case error?
> 3. Does this work well during development? i.e when I'm writing content for a page, and I delete/include icons, will the page reload properly, and will there be any performance issue?

Hi @tlylt,
1. I was planning to create more test pages under test_site_optimize_icons, and a sub-site with no icons at all. I think I forgot about this. Will be adding these soon.
2. I just identified the issue. Cos the MarkBind version in master is updated to v5.0.2 and I have not merged this branch with master. Will be solving this.
3. I have tried some user testing and the pages reload properly and showing the right icon asset files.

**Comment by tlylt** [comment]
> 1. I was planning to create more test pages under test_site_optimize_icons, and a sub-site with no icons at all. I think I forgot about this. Will be adding these soon.

Can we add these in the existing site instead of creating a new one?

**Comment by jmestxr** [comment]
> > 1. I was planning to create more test pages under test_site_optimize_icons, and a sub-site with no icons at all. I think I forgot about this. Will be adding these soon.
> 
> Can we add these in the existing site instead of creating a new one?

The existing test_site contains icons in the site-nav and I plan to create a new site that is clean of icons in the layouts, so I can test the page content independently

**Comment by tlylt** [comment]
> > > 1. I was planning to create more test pages under test_site_optimize_icons, and a sub-site with no icons at all. I think I forgot about this. Will be adding these soon.
> > 
> > 
> > Can we add these in the existing site instead of creating a new one?
> 
> The existing test_site contains icons in the site-nav and I plan to create a new site that is clean of icons in the layouts, so I can test the page content independently

Is testing the page content "independently" absolutely necessary? I assume you can test the conditions that some pages don't have certain icons? If we have to, can we remove the icons in the layouts? does that impact any existing test cases?

Best not to add a new site when there's no need to, for the sake of reducing noise in the test suite.

File: packages/cli/test/functional/test_site/site.json [hunk 1]
```diff
@@ -206,7 +206,7 @@
       "pre": "\n\n# Markbind Plugin Pre-render",
       "post": "<p>Node Modules Plugin Post-render</p>"
     },
-    "filterTags": {
+    "filterTacgs": {
```
Comment (tlylt) [inline packages/cli/test/functional/test_site/site.json:209]: ?

Comment (jmestxr) [reply packages/cli/test/functional/test_site/site.json:209]: Sorry, this must have been a typo. Thanks for pointing this out!

**Comment by tlylt** [comment]
Can we do something about the reduced code coverage? I'm reluctant to review this PR because it :)

**Comment by tlylt** [comment]
Hi @jmestxr, I observe that with this change, pages that don't include certain icon assets will still have the link tag but with empty href attributes, e.g. on https://deploy-preview-2357--markbind-master.netlify.app/
image

Is this a concern? Can we properly remove them? Or are there issues with removing them entirely?

**Comment by jmestxr** [comment]
> Hi @jmestxr, I observe that with this change, pages that don't include certain icon assets will still have the link tag but with empty href attributes, e.g. on https://deploy-preview-2357--markbind-master.netlify.app/ image
> 
> Is this a concern? Can we properly remove them? Or are there issues with removing them entirely?

Hi @tlylt, I have managed to remove the empty <link> tags.

**Comment by jmestxr** [comment]
> Hi @jmestxr, I have not gone into the code. Found some issues with modal (and possibly with the "include" feature?)
> 
> * if you add e.g. `:mio-cloud-download:` into a modal, then opening it will not render the material icon properly, because the css is not included
>   
>   * image
>   * image
>   * strangely Octicons works even without importing the css stylesheet? Need some investigation here
> * I have not tested, but need your help to check if I use an icon in a .md file, and include it in another .md file, will the icon work fine?
> * may also need to check "tab", and "panels"

Hi @tlylt , thanks for catching this. I have updated the code to fix this. In addition, I have tested on modals, panels, quiz. For tabs, this PR isn't the issue that icons could not be rendered.

**Comment by tlylt** [comment]
Hi @jmestxr, I am still seeing issues with modal:
- go to preview https://deploy-preview-2357--markbind-master.netlify.app/userguide/components/popups#modals
- click on 'Slot' to open the modal
image

> For tabs, this PR isn't the issue that icons could not be rendered.

Sorry you mean there is a problem with icon display in tabs right now? 👀, is this reported?

**Comment by jmestxr** [comment]
> Hi @jmestxr, I am still seeing issues with modal:
> 
> * go to preview https://deploy-preview-2357--markbind-master.netlify.app/userguide/components/popups#modals
> * click on 'Slot' to open the modal, see the close button
>   image

@tlylt I did not realise that some vue components also use icons provided by font awesome/glyphicons etc.

I have updated `filterIconAssets` in Page/index.ts which now checks the presence of these icons in both pre vue-rendered HTML and post vue-rendered HTML. Pre-vue HTML does not include the HTML for vue components after rendering, and post-vue does not include HTML of popups like trigger, modals. Hence, need to check both HTML to account for all the icons.

> Sorry you mean there is a problem with icon display in tabs right now? 👀, is this reported?

Yep, I have tried this in the master branch and observed that the slot in the tabs seem to only be processed as text only.

**Comment by tlylt** [comment]
> I have updated filterIconAssets in Page/index.ts which now checks the presence of these icons in both pre vue-rendered HTML and post vue-rendered HTML. Pre-vue HTML does not include the HTML for vue components after rendering, and post-vue does not include HTML of popups like trigger, modals. Hence, need to check both HTML to account for all the icons.

Got it, any performance impact that you observe/foresee for this update?

> Yep, I have tried this in the master branch and observed that the slot in the tabs seem to only be processed as text only, e.g.

Could you help me raise an issue for this?

**Comment by jmestxr** [comment]
> Got it, any performance impact that you observe/foresee for this update?

From what I see, there's no significant change in build time, and time taken to load CSS assets, even for larger icon files like all.min.css (font-awesome styles).

File: packages/cli/test/functional/test_site/expected/testFontAwesomeInPage.html [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        <breadcrumb></breadcrumb>
+
+        <p>
+          <trigger for="modal:fontAwesome">There is one font-awesome icon in this modal</trigger>.
+        </p>
+        <modal id="modal:fontAwesome"><template #header><strong>Modal with font-awesome icon</strong> 🚀</template>
+          <p>A font-awesome icon is supposed to appear here ---&gt; <span aria-hidden="true" class="fas fa-code-branch"></span></p>
+        </modal>
+        <p><strong>There is one font-awesome icon in this page content.</strong></p>
+        <p><strong>Open up console. Only font-awesome stylesheet should be loaded!</strong></p>
```
Comment (tlylt) [inline packages/cli/test/functional/test_site/expected/testFontAwesomeInPage.html:68]: This is not true right? Because line 15-17 imports other icons.

File: packages/cli/test/functional/test_site/testGlyphiconInPage.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<frontmatter>
+title: Test optimize loading of icon stylesheets
+layout: simpleLayoutWithoutIcons.md
+</frontmatter>
+
+<panel header="There are glyphicon icons on this panel."></panel>
+
+**There is are glyphicon icons in this page content.**
```
Comment (tlylt) [inline packages/cli/test/functional/test_site/testGlyphiconInPage.md:8]:
```suggestion
**There are glyphicon icons in this page content.**
```

File: packages/core/src/Page/index.ts [hunk 4]
```diff
@@ -549,6 +576,7 @@ export class Page {
       await this.outputPageHtml(content);
     } else {
       const vueSsrHtml = await pageVueServerRenderer.renderVuePage(compiledVuePage);
+      this.filterIconAssets(content, vueSsrHtml);
```
Comment (tlylt) [inline packages/core/src/Page/index.ts:579]: For the issue of additional imports of other icons in your other test pages, I guess it's either because there are other components such as pageNav/siteNav/layout that contains those icons, or because ` this.filterIconAssets(content, vueSsrHtml);` is not ran for the line 575 ` if (process.env.TEST_MODE)` branch. If it's the latter, we need to adjust where we place this line of coded, and also the functionality of filterIconAssets

What do you think?

File: packages/core/src/Page/index.ts [hunk 5]
```diff
@@ -170,6 +170,33 @@ export class Page {
     };
   }
 
+  /**
+   * Filters out icon asset files that are not used in a page.
+   * @param preVueSsrHtml html content of the page before processing Vue components
```
Comment (tlylt) [inline packages/core/src/Page/index.ts:175]: How about summarize the rationale you gave me in https://github.com/MarkBind/markbind/pull/2357#issuecomment-1752266692 here, so that future devs will know why we are scanning both html content?

**Review by tlylt — CHANGES_REQUESTED** [review]
Hi @jmestxr, thank you for addressing my earlier comments. I think this is mostly ready. I tested locally and seems like it's able to handle edge cases of Modal/Tooltip/include etc.

I left some comments regarding functional test outputs not being accurate according to the description (they still include the other icons), which hopefully will be the last minor issue to tackle.

Lastly, don't forget to update your PR description.

File: packages/core/src/Page/index.ts (hunk 4 above)
Comment (jmestxr) [reply packages/core/src/Page/index.ts:583]: Right, it solves the issue. Thanks for the catch!

Comment (jmestxr) [reply packages/core/src/Page/index.ts:583]: Actually, the stylesheets might not be filtered correctly for TEST_MODE. For example, in testGlyphiconInPage, the actual HTML of <panel> which contains glyphicon icons is not reflected in `content`. As a result, the glyphicon stylesheet is not loaded.

Comment (tlylt) [reply packages/core/src/Page/index.ts:583]: hmm, this is because of vuessr as well? Can we address this by scanning through the compiledVuePage (or something else)?

Comment (tlylt) [reply packages/core/src/Page/index.ts:583]: > Actually, the stylesheets might not be filtered correctly for TEST_MODE

You mean this is a bug?

Comment (jmestxr) [reply packages/core/src/Page/index.ts:583]: it's because TEST_MODE only outputs the pre vue-compiled HTML which consists of "unprocessed" vue components (e.g. `<panel>`,  `<modal>`). The actual HTML of these components might consist of icons. We can simply address this by removing the 'if' condition, i.e. regardless of TEST_MODE or not,  we will output the post vue-compiled HTML.

Comment (tlylt) [reply packages/core/src/Page/index.ts:583]: I see. It's ok I'm fine with the stylesheets not entirely filtered for TEST_MODE.

Comment (tlylt) [reply packages/core/src/Page/index.ts:583]: Hi @jmestxr, sorry I think we should handle the case of filtering for TEST_MODE. We don't have to output the post vue-compiled HTML for TEST_MODE, but we can always render it and give it for filtering. Something like:
```
    const vueSsrHtml = await pageVueServerRenderer.renderVuePage(compiledVuePage);
    this.filterIconAssets(content, vueSsrHtml);
    if (process.env.TEST_MODE) {
      await this.outputPageHtml(content);
    } else {
      await this.outputPageHtml(vueSsrHtml);
    }
```

What do you think?

With this I should expect to see glyphicon CSS import in packages/cli/test/functional/test_site/expected/testGlyphiconInPage.html

Comment (jmestxr) [reply packages/core/src/Page/index.ts]: @tlylt This makes sense and it works. Thanks!

**Review by tlylt — APPROVED** [review]
LGTM, thank you @jmestxr
